### PR TITLE
Jacdac v0

### PR DESCRIPTION
**Pull request by jamesadevine**
This will be the PR for all v0 related jacdac commits (will close previous PRs and consolidate).

### CODE REVIEW

**Comment by jamesadevine** [comment]
- [X] Control Packet refactor
    - [X] devices should broadcast their drivers every 500 ms 
    - [X] as before, control packets should be routed by the logic driver to correct drivers
- [ ] Dynamic packet sizes
    - [ ] device should receive a JDPkt header which will contain the size and the checksum
    - [ ] device should then prepare to receive the remained of the packet
- [ ] Dynamic baud rate
    - [ ] the pulse should dictate the baud rate for communication 10 us is 1mbaud. (i.e. 10 bits at whatever baud)
    - [ ] a device should reconfigure the baud rate upon receiving the pulse and prepare to receive the packet
    - [ ] control packets should be sent at the lowest baud rate on the bus
- [ ] Error detection and recovery
    - [ ] JACDAC should have a dedicated timer for detecting error conditions
    - [ ] As before, bus errors are not permissible in JACDAC
    - [ ] Packet timeouts should be tighter as specified on JACDAC.org
    - [ ] minimum byte gap should be observed.
    - [ ] difference between lo pulse and transmitting baud rate should not be tolerated.
- [ ] JACDAC NACK (for bus level NACKING)

**Comment by jamesadevine** [comment]
Overall this PR is terrible, have another go puny human 😈

File: inc/JACDAC/JACDAC.h [hunk 1]
```diff
@@ -209,7 +217,7 @@ namespace codal
           *
           * @returns DEVICE_OK on success, DEVICE_INVALID_PARAMETER if JD is NULL, or DEVICE_NO_RESOURCES if the queue is full.
           */
-        virtual int send(JDPkt *JD);
+        virtual int send(JDPkt *p);
```
Comment (pelikhan) [inline inc/JACDAC/JACDAC.h]: JD**Pkt** vs JDCommant**Packet**

File: inc/JACDAC/JACDAC.h [hunk 2]
```diff
@@ -251,8 +259,20 @@ namespace codal
...
         JACDACBusState getState();
 
+        /**
+         * Sets the JACDAC baud rate, and internally sets the single wire serial baud JACDAC is using.
+         *
+         * @param baudRate the desired baud rate for this jacdac instance, one of: Baud1M, Baud500K, Baud250K, Baud125K
+         *
+         * @returns DEVICE_OK on success
+         **/
         int setBaud(JACDACBaudRate baudRate);
```
Comment (pelikhan) [inline inc/JACDAC/JACDAC.h]: JDBaudRate

Comment (jamesadevine) [reply inc/JACDAC/JACDAC.h]: Ok!

File: inc/JACDAC/JDProtocol.h [hunk 3]
```diff
@@ -70,69 +70,67 @@ DEALINGS IN THE SOFTWARE.
...
-    struct ControlPacketError
+    struct JDDriverInfo
     {
-        char name[CONTROL_PACKET_ERROR_NAME_LENGTH];
-        int code;
+        uint8_t size;
+        uint8_t address;        // the address assigned by the logic driver
+        uint16_t flags: 8, error_code: 4, type: 4; // various flags, upper eight bits are reserved for control usage, error code, then driver info type
+        uint32_t driver_class;  // the class of the driver
+        uint8_t data[]; // optional additional data
```
[the remark this answers is not in the document]
Comment (pelikhan) [reply inc/JACDAC/JDProtocol.h]: 16bytes sounds good.

File: inc/JACDAC/JACDAC.h (hunk 1 above)
Comment (jamesadevine) [reply inc/JACDAC/JACDAC.h]: I prefer Packet, will rename JDPkt

File: source/JACDAC/JDAccelerometerDriver.cpp [hunk 4]
```diff
@@ -60,12 +60,12 @@ int JDAccelerometerDriver::getZ()
     return latest.z;
 }
 
-int JDAccelerometerDriver::handleControlPacket(JDPkt* cp)
+int JDAccelerometerDriver::handleControlPacket(JDPacket* cp)
```
Comment (jamesadevine) [inline source/JACDAC/JDAccelerometerDriver.cpp]: JDControlPacket

File: source/JACDAC/JACDAC.cpp [hunk 5]
```diff
@@ -50,219 +96,239 @@ uint16_t fletcher16(const uint8_t *data, size_t len)
     return (c1 << 8 | c0);
 }
 
-void JACDAC::dmaComplete(Event evt)
+void jacdac_timer_irq(uint16_t channels)
```
Comment (jamesadevine) [inline source/JACDAC/JACDAC.cpp]: Maybe use a CPP member func?

File: source/JACDAC/JACDAC.cpp [hunk 6]
```diff
@@ -50,219 +96,239 @@ uint16_t fletcher16(const uint8_t *data, size_t len)
...
+
+        return;
+    }
+}
+
+void JACDAC::dmaComplete(Event evt)
+{
+    if (evt.value == SWS_EVT_ERROR)
+    {
+        // event should be more generic: tx/rx timeout
```
Comment (jamesadevine) [inline source/JACDAC/JACDAC.cpp]: Remove comment

File: source/JACDAC/JACDAC.cpp [hunk 7]
```diff
@@ -50,219 +96,239 @@ uint16_t fletcher16(const uint8_t *data, size_t len)
...
+            }
+            else if (status & JD_SERIAL_RECEIVING)
             {
-                // move rxbuf to rxQueue and allocate new buffer.
-                addToQueue(&rxQueue, rxBuf);
-                rxBuf = (JDPkt*)malloc(sizeof(JDPkt));
-                Event(id, JD_SERIAL_EVT_DATA_READY);
-                JD_DMESG("DMA RXD");
+                status &= ~(JD_SERIAL_RECEIVING);
+                system_timer_cancel_event(this->id, JD_SERIAL_EVT_RX_TIMEOUT);
```
Comment (jamesadevine) [inline source/JACDAC/JACDAC.cpp]: Clear interbyte timer instead

File: source/JACDAC/JACDAC.cpp [hunk 8]
```diff
@@ -50,219 +96,239 @@ uint16_t fletcher16(const uint8_t *data, size_t len)
...
-            uint16_t crc = fletcher16(crcPointer, JD_SERIAL_PACKET_SIZE - 2);
+            if (status & JD_SERIAL_RECEIVING_HEADER)
+            {
+
+                status &= ~(JD_SERIAL_RECEIVING_HEADER);
 
-            if (crc == rxBuf->crc)
+                JDPacket* rx = (JDPacket*)rxBuf;
+                sws.receiveDMA(((uint8_t*)rxBuf) + JD_SERIAL_HEADER_SIZE, rx->size);
+                timer.setCompare(MAXIMUM_INTERBYTE_CC, timer.captureCounter() + JD_MAX_INTERBYTE_SPACING);
```
Comment (jamesadevine) [inline source/JACDAC/JACDAC.cpp]: clear MAXIMUM_LO_DATA_CC timer (for Lo -> data  timeout)

File: source/JACDAC/JACDAC.cpp [hunk 9]
```diff
@@ -272,10 +338,15 @@ void JACDAC::initialise()
...
-JACDAC::JACDAC(DMASingleWireSerial&  sws, Pin* busStateLED, Pin* commStateLED, JACDACBaudRate baudRate, uint16_t id) : sws(sws), sp(sws.p), busLED(busStateLED), commLED(commStateLED)
+JACDAC::JACDAC(DMASingleWireSerial&  sws, LowLevelTimer& timer, Pin* busLED, Pin* commStateLED, JDBaudRate baudRate, uint16_t id) : sws(sws), sp(sws.p), timer(timer), busLED(busLED), commLED(commStateLED)
 {
     this->id = id;
-    this->baud = baudRate;
+    this->txBaud = baudRate;
+    instance = this;
+
+    // at least three channels are required.
+    CODAL_ASSERT(timer.getChannelCount() >= TIMER_CHANNELS_REQUIRED);
```
Comment (jamesadevine) [inline source/JACDAC/JACDAC.cpp]: Verify this assert is the right way round

File: source/JACDAC/JACDAC.cpp [hunk 10]
```diff
@@ -343,82 +400,149 @@ void JACDAC::stop()
...
+    bool toggle = false;
+    if (e.value == DEVICE_PIN_EVT_RISE)
     {
-        JD_DMESG("WAIT %s", (status & JD_SERIAL_BUS_RISE) ? "LO" : "REC");
-        if (status & JD_SERIAL_BUS_RISE)
+        // Lo to hi transition?
+        if (status & JD_SERIAL_BUS_LO)
+            toggle = true;
+
+        status |= (JD_SERIAL_BUS_HI | ((toggle) ? JD_SERIAL_BUS_TOGGLED : 0));
```
Comment (jamesadevine) [inline source/JACDAC/JACDAC.cpp]: Mask off the alternate bit (or maybe just condense into a single bit?

File: source/JACDAC/JACDAC.cpp [hunk 11]
```diff
@@ -109,28 +169,49 @@ void JACDAC::dmaComplete(Event evt)
...
     sp.setDigitalValue(1);
-    configure(true);
+    configure(JACDACPinEvents::PulseEvents);
 
     if (commLED)
         commLED->setDigitalValue(0);
 }
 
-void JACDAC::onFallingEdge(Event)
+void JACDAC::onLowPulse(Event e)
```
[the remark this answers is not in the document]
Comment (jamesadevine) [reply source/JACDAC/JACDAC.cpp]: Need to add a class that directly calls a function pointer (override pin class)

**Comment by jamesadevine** [comment]
- [ ] EUI 64

File: inc/JACDAC/services/JDConsoleService.h [hunk 12]
```diff
@@ -4,26 +4,26 @@
 #include "JACDAC.h"
 #include "ManagedString.h"
 
-#define JD_CONSOLE_LOG_PRIORITY_LOG         0
-#define JD_CONSOLE_LOG_PRIORITY_ERROR       1
-#define JD_CONSOLE_LOG_PRIORITY_WARNING     2
+#define JD_CONSOLE_LOG_PRIORITY_LOG         1
+#define JD_CONSOLE_LOG_PRIORITY_INFO        2
 #define JD_CONSOLE_LOG_PRIORITY_DEBUG       3
+#define JD_CONSOLE_LOG_PRIORITY_ERROR       4
```
Comment (pelikhan) [inline inc/JACDAC/services/JDConsoleService.h]: Are you sure about the order? debug < log < warn < error

Comment (jamesadevine) [reply inc/JACDAC/services/JDConsoleService.h]: Is that the order? happy to modify

File: source/drivers/KeyValueStorage.cpp [hunk 13]
```diff
@@ -110,7 +110,7 @@ void KeyValueStorage::scratchKeyValueStore(KeyValueStore store)
   */
 void KeyValueStorage::scratchKeyValuePair(KeyValuePair pair, int scratchOffset)
 {
-    memcpy(this->scratch + scratchOffset, &pair, sizeof(KeyValueStore));
+    memcpy(this->scratch + scratchOffset, &pair, sizeof(KeyValuePair));
```
Comment (pelikhan) [inline source/drivers/KeyValueStorage.cpp]: funny.

Comment (jamesadevine) [reply source/drivers/KeyValueStorage.cpp]: This took me at least 4 hours to track down.

Comment (jamesadevine) [reply source/drivers/KeyValueStorage.cpp]: 😠 Careless 12am code!

**Pull request merged**